# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


url = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/titanic_train.csv'
titanic_series = pd.read_csv(url, usecols=["Name"]).squeeze("columns")

In [2]:
import pandas as pd

In [3]:
url = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv'
insurance_df = pd.read_csv(url)

In [4]:
url = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv'
insurance_series = pd.read_csv(url, usecols=["Customer"]).squeeze("columns")

In [5]:
insurance_df.head(5)

Customer          ST GENDER             Education Customer Lifetime Value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    Income  Monthly Premium Auto Number of Open Complaints     Policy Type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   Vehicle Class  Total Claim Amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- 1) Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- 2) Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- 3) Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- 4) Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- 5) Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [6]:
## Your code here
## Ex 1.1

insurance_df.shape

## There are 4008 rows and 11 columns

(4008, 11)

In [7]:
## Ex 1.2, Part A

**Customer** -> Unique ID of Customer; categorical data stored stored as Object(String)

**ST** -> American State of Residence, of Customer; categorical data stored as Object(String)

**GENDER** -> Gender of Customer; categorical data stored as Object(String)

**Education** -> Highest degree of education of Customer; categorical data stored as Object(String)

**Customer Lifetime Value** -> Amount of Gewinn made on said Customer, in percentage; numerical data apparently stored as Object(String)

**Income** -> Yearly Income of Customer; numerical data stored as Float

**Monthly Premium Auto** -> Monthly Fee acquired from the Customer; numerical data saved as stored as Float

**Number of Open Complaints** -> Amount of Open Complaints of a said Customer; numerical data saved as Object(String)

**Vehicle Class** -> The Category of Product(Car) used by the Customer; categorical data saved as Object(String)

**Total Claim Amount** -> The value sum of all claims of a Customer; numerical data save as Float

In [8]:
## Ex 1.2, Part B

**Customer** -> Contains null values, they should be replaced with dummy ID

**ST** -> There is no convention as to the naming, therefore some states are stated multiple times. Example: California and Cali. We could identify all potential variants and transform them into California. This could be done with a loop.

**GENDER** -> As with the previous category, the entries have been done in many variant ways. We should identify and aggregate them.

**Education** -> A few duplicate categories and null values. The former can be aggreg, and the rest ignored.

**Customer Lifetime Value** -> Due to the percentage sign, these seem to be unfortunely stored as Object(String), making an assessment of them impsosible without transforming the data first. We could slice the % values of and then divide them by 100 to get a compatbile float, which could interact with the other numerical data, at which point it would be move to the numerical_variables group, as it should.

**All Numerical Data** -> Only a fraction (1/4) of the data seems to have non-null entries, which heavely skews the data due to the blindspot. At this point it is the best to leave the 3/4 out of it and analyse what we have.

**Income** -> There seem to be a significant amount of entries equal to 0 or otherwise very low, which are suspicious. These should examined

**Vehicle Class** -> A few nulls, otherwise okay.

**Number of Open Complaints** -> Seems to be in a specific data format, which unfortunely would have to be transformed into integers before they can be better handled. However, despite its state, it can already be read in a limited faction. To solve this, we could see the pattern of this format, and slice the data that we want out of it . In this case, any integer between the /, as in YY/x/YY.


In [9]:
## Ex 1.3

insurance_df.nunique()

Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64

In [10]:
## Categorical Data -> Customer, ST, GENDER, Education, Policy Type, Vehicle Class

categorical_variables = ["Customer","ST","GENDER","Education","Customer Lifetime Value","Policy Type", "Vehicle Class", "Number of Open Complaints"]
insurance_df[categorical_variables]

Customer          ST GENDER             Education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
4003      NaN         NaN    NaN                   NaN   
4004      NaN         NaN    NaN                   NaN   
4005      NaN         NaN    NaN                   NaN   
4006      NaN         NaN    NaN                   NaN   
4007      NaN         NaN    NaN                   NaN   

     Customer Lifetime Value     Policy Type  Vehicle Class  \
0                        NaN   Personal Auto  Four-Door Car   
1                 697953.59%   Personal Auto  Four-Door Car   
2                1288743.17%   Personal Auto   Two-Door Car   
3                 764586.18%  Corporate Auto            SUV   
4                 536307.65%   Personal Auto  Four-Door Car   
...                      ...             ...            ...   
4003                     NaN             NaN            NaN   
4004                     NaN             NaN            NaN   
4005                     NaN             NaN            NaN   
4006                     NaN             NaN            NaN   
4007                     NaN             NaN            NaN   

     Number of Open Complaints  
0                       1/0/00  
1                       1/0/00  
2                       1/0/00  
3                       1/0/00  
4                       1/0/00  
...                        ...  
4003                       NaN  
4004                       NaN  
4005                       NaN  
4006                       NaN  
4007                       NaN  

[4008 rows x 8 columns]

In [11]:
#insurance_df[categorical_variables].unique()

insurance_df.Customer.unique()

array(['RB50392', 'QZ44356', 'AI49188', ..., 'CW49887', 'MY31220', nan],
      dtype=object)

In [12]:
## Getting unique values of all Categorical Columns

for i in categorical_variables:
    print(f"Start of unique values of {i}:")
    print(insurance_df[i].unique())
    print(f"---------------")

Start of unique values of Customer:
['RB50392' 'QZ44356' 'AI49188' ... 'CW49887' 'MY31220' nan]
---------------
Start of unique values of ST:
['Washington' 'Arizona' 'Nevada' 'California' 'Oregon' 'Cali' 'AZ' 'WA'
 nan]
---------------
Start of unique values of GENDER:
[nan 'F' 'M' 'Femal' 'Male' 'female']
---------------
Start of unique values of Education:
['Master' 'Bachelor' 'High School or Below' 'College' 'Bachelors' 'Doctor'
 nan]
---------------
Start of unique values of Customer Lifetime Value:
[nan '697953.59%' '1288743.17%' ... '2031499.76%' '323912.47%'
 '899704.02%']
---------------
Start of unique values of Policy Type:
['Personal Auto' 'Corporate Auto' 'Special Auto' nan]
---------------
Start of unique values of Vehicle Class:
['Four-Door Car' 'Two-Door Car' 'SUV' 'Luxury SUV' 'Sports Car'
 'Luxury Car' nan]
---------------
Start of unique values of Number of Open Complaints:
['1/0/00' '1/2/00' '1/1/00' '1/3/00' '1/5/00' '1/4/00' nan]
---------------


In [13]:
# Display unique values and their frequency counts for each categorical variable
for j in categorical_variables:
    print(f"Start of amount per value of {j}:")
    print(insurance_df[j].value_counts())
    print(f"---------------")

Start of amount per value of Customer:
Customer
RB50392    1
HJ15383    1
AO74776    1
HQ82233    1
OL72737    1
          ..
RO26085    1
ES57969    1
JK55587    1
RN97635    1
MY31220    1
Name: count, Length: 1071, dtype: int64
---------------
Start of amount per value of ST:
ST
Oregon        320
California    211
Arizona       186
Cali          120
Nevada         98
Washington     81
WA             30
AZ             25
Name: count, dtype: int64
---------------
Start of amount per value of GENDER:
GENDER
F         457
M         413
Male       39
female     28
Femal      17
Name: count, dtype: int64
---------------
Start of amount per value of Education:
Education
Bachelor                324
College                 313
High School or Below    296
Master                   94
Doctor                   37
Bachelors                 7
Name: count, dtype: int64
---------------
Start of amount per value of Customer Lifetime Value:
Customer Lifetime Value
445811.34%     4
251459.20%     4
272

In [14]:
numerical_variables = ["Income", "Monthly Premium Auto", "Total Claim Amount"]
insurance_df[numerical_variables]

Income  Monthly Premium Auto  Total Claim Amount
0         0.0                1000.0            2.704934
1         0.0                  94.0         1131.464935
2     48767.0                 108.0          566.472247
3         0.0                 106.0          529.881344
4     36357.0                  68.0           17.269323
...       ...                   ...                 ...
4003      NaN                   NaN                 NaN
4004      NaN                   NaN                 NaN
4005      NaN                   NaN                 NaN
4006      NaN                   NaN                 NaN
4007      NaN                   NaN                 NaN

[4008 rows x 3 columns]

In [15]:
## Ex 1.3 continued

insurance_df[numerical_variables].describe()

### MEAN = AVERAGE

Income  Monthly Premium Auto  Total Claim Amount
count   1071.000000           1071.000000         1071.000000
mean   39295.701214            193.234360          404.986909
std    30469.427060           1601.190369          293.027260
min        0.000000             61.000000            0.382107
25%    14072.000000             68.000000          202.157702
50%    36234.000000             83.000000          354.729129
75%    64631.000000            109.500000          532.800000
max    99960.000000          35354.000000         2893.239678

In [16]:
for h in numerical_variables:
    print(f"Start of amount per value of {h}:")
    print(insurance_df[h].value_counts())
    print(f"---------------")

Start of amount per value of Income:
Income
0.0        252
43860.0      4
17622.0      4
36650.0      3
14072.0      3
          ... 
29486.0      1
58557.0      1
25632.0      1
18768.0      1
54230.0      1
Name: count, Length: 774, dtype: int64
---------------
Start of amount per value of Monthly Premium Auto:
Monthly Premium Auto
65.0     47
63.0     39
71.0     37
67.0     37
73.0     37
         ..
177.0     1
181.0     1
276.0     1
253.0     1
199.0     1
Name: count, Length: 132, dtype: int64
---------------
Start of amount per value of Total Claim Amount:
Total Claim Amount
321.600000    19
331.200000    15
312.000000    14
350.400000    14
316.800000    12
              ..
421.484456     1
430.505942     1
7.646763       1
55.510526      1
207.320041     1
Name: count, Length: 761, dtype: int64
---------------


- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.

In [17]:
## Ex 1.4

- The average income of a customer is 39.294,70 €, with the median being 36.234,00 €
- Half of the customer had made at least 532.80 € (median) out of their claims.


In [18]:
## Ex 1.5

a

## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [19]:
# Your code here

insurance_df["ST"].value_counts()

## insurance_df["ST"].value_counts().tail(5)

ST
Oregon        320
California    211
Arizona       186
Cali          120
Nevada         98
Washington     81
WA             30
AZ             25
Name: count, dtype: int64

In [20]:
d1 = {"state": ["Oregon", "California", "Arizona", "Nevada", "Washington"],
     "count": [320,211+120,186+25,98,81+30]
}

mkt_d1 = pd.DataFrame(d1)
mkt_d1

state  count
0      Oregon    320
1  California    331
2     Arizona    211
3      Nevada     98
4  Washington    111

In [21]:
mkt_d1.sort_values(by="count", ascending=True)

state  count
3      Nevada     98
4  Washington    111
2     Arizona    211
0      Oregon    320
1  California    331

In [22]:
a = 0
for i in mkt_d1["count"]:
    a+=i
print(a)
for p in mkt_d1["count"]:
    print(f"Percentage of {p} is {p/a*100}")

1071
Percentage of 320 is 29.878618113912232
Percentage of 331 is 30.905695611577965
Percentage of 211 is 19.70121381886088
Percentage of 98 is 9.15032679738562
Percentage of 111 is 10.364145658263306


In [23]:
mkt_d1["count"][0]

320

Hello marketing team, please observe our findings below. Please take into consideration that we have taken the liberty to clean up the data and merge duplicates.

States by least amont of entries:

- Nevada, 98 customers (9.2%)
- Washington, 111 customers (10.4%)
- Arizona, 211 customers (19.7%)
- Oregon, 320 customers (29.9%)
- California, 331 (30.9%)

### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [24]:
# Your code here

insurance_df["Policy Type"].value_counts()


Policy Type
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: count, dtype: int64

In [25]:
insurance_df["Policy Type"].describe()

count              1071
unique                3
top       Personal Auto
freq                780
Name: Policy Type, dtype: object

### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [37]:
# Your code here

ex3 = ["Income", "Policy Type"]

insurance_df[ex3].value_counts().head(10)



Income   Policy Type   
0.0      Personal Auto     189
         Corporate Auto     55
         Special Auto        8
17622.0  Personal Auto       4
20396.0  Personal Auto       2
95697.0  Personal Auto       2
28142.0  Personal Auto       2
74454.0  Personal Auto       2
67763.0  Personal Auto       2
26049.0  Personal Auto       2
Name: count, dtype: int64

In [58]:
insurance_df.filter(items=["Income", "Policy Type"=="Personal Auto"]).mean()

Income    39295.701214
dtype: float64

In [61]:
insurance_df.filter(items=["Income", "Policy Type"=="Corporate Auto"]).mean()

Income    39295.701214
dtype: float64

In [65]:
ex3_personal_filter_df = insurance_df[insurance_df["Policy Type"] == "Personal Auto"]

In [78]:
personal_auto_income_mean = ex3_personal_filter_df["Income"].mean().round(1)
print(personal_auto_income_mean)

38180.7


In [79]:
ex3_corporate_filter_df = insurance_df[insurance_df["Policy Type"] == "Corporate Auto"]

In [80]:
corporate_auto_income_mean = ex3_corporate_filter_df["Income"].mean().round(1)
print(corporate_auto_income_mean)

41390.3


In [81]:
print(f"Hello Sales team: The average income of customers with the Personal Auto policy is {personal_auto_income_mean}, while the average income of customers with the Corporate Auto policy is {corporate_auto_income_mean}")

Hello Sales team: The average income of customers with the Personal Auto policy is 38180.7, while the average income of customers with the Corporate Auto policy is 41390.3


In [54]:
#insurance_df[::1][ex3]

Income     Policy Type
0         0.0   Personal Auto
1         0.0   Personal Auto
2     48767.0   Personal Auto
3         0.0  Corporate Auto
4     36357.0   Personal Auto
...       ...             ...
4003      NaN             NaN
4004      NaN             NaN
4005      NaN             NaN
4006      NaN             NaN
4007      NaN             NaN

[4008 rows x 2 columns]

In [82]:
#insurance_df.loc[0][0]

In [83]:
#insurance_df["Policy Type"].describe(include="Personal Auto")

In [84]:
#insurance_df["Policy Type"].Income.mean()

### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [ ]:
# Your code here